x/programs/rust/examples/multisig/src/lib.rs

```diff
@@ -0,0 +1,495 @@
+    }
+
+    context
+        .store([
+            (StateKeys::Nays(last_id), &0u64),
+            (StateKeys::Yeas(last_id), &1),
+        ])
+        .unwrap();
+
+    context
+        .store_by_key(StateKeys::LastProposalId, &(last_id + 1))
+        .unwrap();
+
+    last_id
+}
+
+#[public]
+pub fn vote(context: Context<StateKeys>, id: u64, yea: bool) -> Result<(), ProposalError> {
+    let proposal = _proposal_at(&context, id).ok_or(ProposalError::InexistentProposal)?;
+
+    if proposal.executed {
+        return Err(ProposalError::AlreadyExecuted);
+    }
+
+    let actor = context.actor();
+    let voters: Vec<_> = (0..proposal.voters_len)
+        .map(|i| {
+            context
+                .get(StateKeys::Voter(id, i as usize))
+                .unwrap()
+                .unwrap()
+        })
+        .collect();
+
+    let i = _ensure_voter(&actor, &voters)?;
+
+    if yea {
+        context
+            .store_by_key(
+                StateKeys::Yeas(id),
+                &(context.get::<u64>(StateKeys::Yeas(id)).unwrap().unwrap() + 1),
+            )
+            .unwrap();
+    } else {
+        context
+            .store_by_key(
+                StateKeys::Nays(id),
+                &(context.get::<u64>(StateKeys::Nays(id)).unwrap().unwrap() + 1),
+            )
+            .unwrap();
+    }
+
+    context
+        .store_by_key(
+            StateKeys::Voter(id, i as usize),
+            &Voter {
+                address: actor,
+                voted: true,
+            },
+        )
+        .unwrap();
+
+    Ok(())
+}
+
+#[public]
+pub fn execute(
+    context: Context<StateKeys>,
+    proposal_id: u64,
+    max_units: u64,
+) -> Result<DeferDeserialize, ProposalError> {
+    let mut proposal =
+        _proposal_at(&context, proposal_id).ok_or(ProposalError::InexistentProposal)?;
+
+    if proposal.executed {
+        return Err(ProposalError::AlreadyExecuted);
+    }
+
+    let yeas = context.get(StateKeys::Yeas(proposal_id)).unwrap().unwrap();
+    let nays = context.get(StateKeys::Nays(proposal_id)).unwrap().unwrap();
+    if !_quorum_reached(proposal.voters_len, yeas, nays) {
+        return Err(ProposalError::QuorumNotReached);
+    }
+
+    proposal.executed = true;
+
+    proposal
+        .to
+        .call_function::<DeferDeserialize>(
+            &proposal.method,
+            &proposal.args,
+            max_units,
+            proposal.value,
+        )
+        .map_err(ProposalError::ExecutionFailed)
+}
+
+#[public]
+pub fn last_proposal_id(context: Context<StateKeys>) -> u64 {
+    _last_proposal_id(&context)
+}
+
+#[public]
+pub fn quorum_reached(
+    context: Context<StateKeys>,
+    proposal_id: u64,
+) -> Result<bool, ProposalError> {
+    let proposal = _proposal_at(&context, proposal_id).ok_or(ProposalError::InexistentProposal)?;
+    let yeas = context.get(StateKeys::Yeas(proposal_id)).unwrap().unwrap();
+    let nays = context.get(StateKeys::Nays(proposal_id)).unwrap().unwrap();
```

Comment:
You should be using an `expect` with a `unwrap_or_default` here (and elsewhere)

Reply:
I disagree with the use of a `unwrap_or_default` here, because if the "default" case is returned then an important invariant is broken. I have instead used two `expect` since this function should be called after a proposal has been created. The state should be correctly initialized after the call to `propose`.